### ISSUE TITLE
Require lsp-stdio-connection to take a function

### ISSUE THREAD

**Issue by leungbk** [emacs-lsp/lsp-mode#1993]
There are a few clients that expose a `(defcustom lsp-foo-server-executable-path "some-server-bin")` while also registering the client with `lsp-client-new-connection` set to `(lsp-stdio-connection lsp-foo-server-executable-path)`. This is a bit confusing since a user who sees and modifies the `defcustom` may find that the executable used by `lsp-mode` is the same unless they run `lsp-register-client` again.

Other clients wrap the executable path with a function, which will not have this problem. So I think it might be good to make this the standard.

**Comment by yyoncho** [comment]
That is true - we should make that consistent. IMO the way to go is to generate properties from lsp-register-client. Your comment applies to every single setting passed to lsp-register-client. What @kiennq suggested is to turn lsp-register-client into macro. But having it as a macro will cause update issues when we change it and an external package is bytecompiled against the old one.

**Comment by leungbk** [comment]
> But having it as a macro will cause update issues when we change it and an external package is bytecompiled against the old one.

Can you describe how to reproduce this? After installing from Melpa, I can install `lsp-mode` from source with `package-install-file` after having uncommented https://github.com/emacs-lsp/lsp-mode/blob/a8848b0e8292d69b745251d746b9492eb4e4c22f/lsp-eslint.el#L219 It appears to install properly, so that tells me I am doing something wrong.

`helm` and `magit` have had similar issues, so I submitted https://github.com/jwiegley/emacs-async/pull/128 to try and fix this.

**Comment by yyoncho** [comment]
Here are the steps to reproduce:

Create a new file `boo.el` and insert 

```
(defun bar ()
  (foo-bar))
```

Then go in this file and do `M-x emacs-lisp-byte-compile-and-load`

then eval
```
(defmacro foo-bar () )
```

After that, if you eval `(bar)` it will fail with invalid function if you define foo-bar as function it will work.

**Comment by kiennq** [comment]
Come back to this. I think we can just issue a warning when `lsp-stdio-connection` is used with a symbol instead of function.
That will happens once and make sure both user/maintainer aware of what should be done
